habito/habito.py

```diff
@@ -103,9 +115,18 @@ def list():
 
             if quanta >= habit.quantum:
                 column_text = u'\u2713'
-
-            habit_row.append("{0} ({1})".format(column_text, quanta))
-
+                if in_streak:
```

Comment:
I'd suggest to start with just the minimum refactoring required to separate
the responsibility and get going with the features. May be it will evolve
to the complex cli pattern in near future, we will have more clarity on the
design then to decide.

`list` will be the most used command, given other commands take `id` as
input. It still feels this operation will slow down if it involves
computation, with check-in, the computation of streak is additive if user
already has a streak, if not just start a new streak. If user needs
additional data, e.g. # of streaks in past month, we will compute it on the
fly because user asked for it :)

Yes, achievements being non-continuous may not be the right place for
streaks. Are there are similar *continuous* data that needs to be
maintained per habit? E.g. an overall goal, progress towards the goal, a
proposed quanta required to meet the goal etc.? Just brainstorming to see
if `streak` could fit into such a model.

On Sun, May 22, 2016 at 4:01 PM, Ben Baert <notifications@github.com> wrote:

> In habito/habito.py
> <https://github.com/codito/habito/pull/3#discussion_r64147756>:
>
> > @@ -103,9 +115,18 @@ def list():
> >
> >              if quanta >= habit.quantum:
> >                  column_text = u'\u2713'
> > -
> > -            habit_row.append("{0} ({1})".format(column_text, quanta))
> > -
> > +                if in_streak:
>
> As for the complexity issue you raised, I definitely agree. I think we
> should apply the 'complex pattern', as shown here,
> https://github.com/pallets/click/tree/master/examples/complex, and also
> have a models and utilities file. From there on, we can start building
> these new features in a more sensible manner and start separating the view
> and the model, as you suggested. If you agree, I'm definitely willing to
> work on this next week.
>
> I think it's best for streak to be calculated each time, since it can
> change so often. It's definitely possible to do this more efficiently I do
> like the AchievementsModel suggestion though, but it would be more useful
> for badges etc. (e.g. 5 checkins, 10 checkins, 5-day streak (at least
> once), ...).
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/codito/habito/pull/3/files/8fbb089e624977c99f2258df7a9c948eb2d8c13c#r64147756>
>